core/src/main/java/io/grpc/GrpcSession.java

```diff
@@ -0,0 +1,66 @@
+package io.grpc;
+
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.SSLSession;
+import javax.security.cert.X509Certificate;
+import java.net.SocketAddress;
+
+/**
+ * Holds state pertaining to a single transport connection
+ */
+public class GrpcSession {
+
+  private final SocketAddress remoteAddress;
+  private SSLEngine sslEngine;
+
+  public GrpcSession(SocketAddress remoteAddress) {
+    this.remoteAddress = remoteAddress;
+  }
+
+  public SocketAddress getRemoteAddress() {
+    return remoteAddress;
+  }
+
+  public SSLSession getSslSession() {
```

Comment:
This may be exposing too much as SSLSession is mutable and it isn't a given we will be using SSL/TLS. It seems we only need to provide a Principal (sslEngine.getSession().getPeerPrincipal()).